src/ConnectedMode/QualityProfiles/QualityProfileDownloader.cs

```diff
@@ -0,0 +1,202 @@
+        private readonly IConfigurationPersister configurationPersister;
+        private readonly ISolutionBindingOperation solutionBindingOperation;
+
+        private readonly ILogger logger;
+
+        private readonly IEnumerable<Language> languagesToBind;
+
+        [ImportingConstructor]
+        public QualityProfileDownloader(
+            ISonarQubeService sonarQubeService,
+            IBindingConfigProvider bindingConfigProvider,
+            IConfigurationPersister configurationPersister,
+            ISolutionBindingOperation solutionBindingOperation,
+            ILogger logger) :
+            this(
+                sonarQubeService,
+                bindingConfigProvider,
+                configurationPersister,
+                solutionBindingOperation,
+                logger,
+                languagesToBind: Language.KnownLanguages
+                )
+        { }
+
+        internal /* for testing */ QualityProfileDownloader(
+            ISonarQubeService sonarQubeService,
+            IBindingConfigProvider bindingConfigProvider,
+            IConfigurationPersister configurationPersister,
+            ISolutionBindingOperation solutionBindingOperation,
+            ILogger logger,
+            IEnumerable<Language> languagesToBind)
+        {
+            this.bindingConfigProvider = bindingConfigProvider;
+            this.sonarQubeService = sonarQubeService;
+            this.configurationPersister = configurationPersister;
+            this.solutionBindingOperation = solutionBindingOperation;
+            this.logger = logger;
+            this.languagesToBind = languagesToBind;
+        }
+
+        public async Task<bool> UpdateAsync(BoundSonarQubeProject boundProject, IProgress<FixedStepsProgress> progress, CancellationToken cancellationToken)
+        {
+            // TODO - CancellableJobRunner
+            // TODO - threading
+            // TODO - skip downloading up to date QPs
+
+            EnsureProfilesExistForAllSupportedLanguages(boundProject);
+
+            var bindingConfigs = new List<IBindingConfig>();
+
+            var languageCount = languagesToBind.Count();
+            int currentLanguage = 0;
+
+            foreach (var language in languagesToBind)
+            {
+                currentLanguage++;
+
+                var progressMessage = string.Format(BindingStrings.DownloadingQualityProfileProgressMessage, language.Name);
+                progress?.Report(new FixedStepsProgress(progressMessage, currentLanguage, languageCount));
+
+                var qualityProfileInfo = await TryDownloadQualityProfileAsync(boundProject, language, cancellationToken);
+
+                if (qualityProfileInfo == null)
+                {
+                    continue; // skip to the next language
+                }
+                UpdateProfile(boundProject, language, qualityProfileInfo);
+
+                var bindingConfiguration = configurationPersister.Persist(boundProject);
```

Comment:
Lines 114 and 115 do what the old `QueueWriteBindingInformation` method did i.e. update the new server profile key and timestamp and write the information to the `binding.config` file.